cmd/snapshot_export.go

```
@@ -17,38 +27,175 @@ func (o *snapshotExportOutput) toJSON()  { outputJSON(o) }
-			if len(args) != 1 {
-				return cmd.Usage()
-			}
-			return output(exportSnapshot(args[0]))
-		},
-	})
+		strings.Join(outputterTemplateAnnotations(&snapshotExportOutput{}), ", ")),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) != 1 {
+			return cmd.Usage()
+		}
+
+		filePath, err := cmd.Flags().GetString("download")
+		if err != nil {
+			return err
+		}
+
+		snapshot, err := exportSnapshot(args[0])
+		if err != nil {
+			return err
+		}
+
+		if !cmd.Flags().Changed("download") {
+			return output(&snapshotExportOutput{
+				URL:      snapshot.PresignedURL,
+				Checksum: snapshot.MD5sum,
+			}, nil)
+		}
+
+		filePath, err = downloadExportedSnapshot(filePath, args[0], snapshot.PresignedURL)
+		if err != nil {
+			return err
+		}
+
+		fmt.Print("Verifying downloaded file checksum... ")
+
+		if err = checkExportedSnapshot(filePath, snapshot.MD5sum); err != nil {
+			return err
+		}
+
+		fmt.Println("success")
+
+		return nil
+	},
 }
 
-func exportSnapshot(snapshotID string) (outputter, error) {
+func exportSnapshot(snapshotID string) (*egoscale.ExportSnapshotResponse, error) {
 	id, err := egoscale.ParseUUID(snapshotID)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("exporting snapshot %q", id))
+	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("Exporting snapshot %q", id))
 	if err != nil {
 		return nil, err
 	}
 
-	snapshot := res.(*egoscale.ExportSnapshotResponse)
+	return res.(*egoscale.ExportSnapshotResponse), nil
+}
+
+func downloadExportedSnapshot(filePath, snapshotID, url string) (string, error) {
+	filePath = filepath.ToSlash(filePath)
+
+	st, err := os.Stat(filePath)
+	if err == nil {
+		if st.IsDir() {
+			filePath = filepath.Join(filePath, snapshotID+".qcow2")
+			_, err = os.Stat(filePath)
+			if !os.IsNotExist(err) {
+				return "", fmt.Errorf("file %q: already exists", filePath)
+			}
+		}
```

Here if the path is not a dir, I think you can return the same error as above 
```Golang
return "", fmt.Errorf("file %q: already exists", filePath)
```